### PR TITLE
sql: fix race in TestPGWireConnectionCloseReleasesLeases

### LINKED ISSUE #7750: circleci: failed tests: TestPGWireConnectionCloseReleasesLeases
The following test appears to have failed:

[#20112](https://circleci.com/gh/cockroachdb/cockroach/20112):

```
W160711 20:52:35.081090 kv/dist_sender.go:776  key range /Table/51/1-/Table/51/2 outside of bounds of range /Table/14-/Table/50
I160711 20:52:35.083139 storage/replica.go:1862  applying command with forced error: storage/replica.go:1783: no-op on empty Raft entry
E160711 20:52:35.083770 storage/replica.go:1873  [node=1,store=1,range=6] error executing raft command: storage/replica.go:1783: no-op on empty Raft entry
I160711 20:52:35.084103 storage/replica_command.go:1630  range 6: new range lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 407852h52m39.632243287s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-07-11 20:52:35.083980985 +0000 UTC]
==================
WARNING: DATA RACE
Write by goroutine 16:
  github.com/cockroachdb/cockroach/sql.(*tableState).release.func1()
      /go/src/github.com/cockroachdb/cockroach/sql/lease.go:688 +0x165
  github.com/cockroachdb/cockroach/sql.(*tableState).release()
      /go/src/github.com/cockroachdb/cockroach/sql/lease.go:700 +0x3e9
  github.com/cockroachdb/cockroach/sql.(*LeaseManager).Release()
      /go/src/github.com/cockroachdb/cockroach/sql/lease.go:1094 +0x254
  github.com/cockroachdb/cockroach/sql.(*planner).releaseLeases()
      /go/src/github.com/cockroachdb/cockroach/sql/table.go:362 +0x170
  github.com/cockroachdb/cockroach/sql.(*Session).Finish()
--
  github.com/cockroachdb/cockroach/server.(*Server).Start.func8.1()
      /go/src/github.com/cockroachdb/cockroach/server/server.go:370 +0x75
  github.com/cockroachdb/cockroach/util/netutil.(*Server).ServeWith.func1()
      /go/src/github.com/cockroachdb/cockroach/util/netutil/net.go:131 +0x93

Previous read by goroutine 47:
  github.com/cockroachdb/cockroach/sql.TestPGWireConnectionCloseReleasesLeases.func1()
      /go/src/github.com/cockroachdb/cockroach/sql/pgwire_internal_test.go:81 +0xcf
  github.com/cockroachdb/cockroach/util.RetryForDuration()
      /go/src/github.com/cockroachdb/cockroach/util/testing.go:125 +0x192
  github.com/cockroachdb/cockroach/util.SucceedsSoonDepth()
      /go/src/github.com/cockroachdb/cockroach/util/testing.go:112 +0x44
  github.com/cockroachdb/cockroach/util.SucceedsSoon()
      /go/src/github.com/cockroachdb/cockroach/util/testing.go:106 +0x4b
  github.com/cockroachdb/cockroach/sql.TestPGWireConnectionCloseReleasesLeases()
      /go/src/github.com/cockroachdb/cockroach/sql/pgwire_internal_test.go:86 +0xf03
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:473 +0xdc

Goroutine 16 (running) created at:
  github.com/cockroachdb/cockroach/util/netutil.(*Server).ServeWith()
      /go/src/github.com/cockroachdb/cockroach/util/netutil/net.go:133 +0x397
  github.com/cockroachdb/cockroach/server.(*Server).Start.func8()
      /go/src/github.com/cockroachdb/cockroach/server/server.go:373 +0xd4
  github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1()
      /go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x5f

Goroutine 47 (running) created at:
  testing.RunTests()
      /usr/local/go/src/testing/testing.go:582 +0xae2
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:515 +0x11d
  github.com/cockroachdb/cockroach/sql_test.TestMain()
      /go/src/github.com/cockroachdb/cockroach/sql/main_test.go:197 +0xf8
  main.main()
      github.com/cockroachdb/cockroach/sql/_test/_testmain.go:602 +0x20a
==================
I160711 20:52:35.090483 storage/engine/rocksdb.go:333  closing in-memory rocksdb instance
--
1      kv/txn_coord_sender.go:873
I160711 20:55:06.892077 storage/engine/rocksdb.go:333  closing in-memory rocksdb instance
--- PASS: TestReacquireLeaseOnRestart (0.72s)
PASS
Found 1 data race(s)
FAIL	github.com/cockroachdb/cockroach/sql	161.331s
I160711 20:52:16.561094 rand.go:76  Random seed: -2794217228725600103
=== RUN   TestProcessExpression
--- PASS: TestProcessExpression (0.01s)
=== RUN   TestFlowRegistry
--- PASS: TestFlowRegistry (0.30s)
=== RUN   TestOrderedSync
--- PASS: TestOrderedSync (0.00s)
=== RUN   TestJoinReader
I160711 20:52:16.882037 storage/engine/rocksdb.go:289  opening in memory rocksdb instance
W160711 20:52:16.913712 gossip/gossip.go:934  not connected to cluster; use --join to specify a connected node

```
Please assign, take a look and update the issue accordingly.

### PR DESCRIPTION
**Pull request by andreimatei**
Fixes #7750

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/7754)

### CODE REVIEW

**Comment by tamird** [comment]
LGTM but slightly prefer
```
m.Lock()
val := ...
m.Unlock()
```
to the current patterns

**Comment by andreimatei** [comment]
done

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7754#-:-KMQuunBMzXYrFbeF7vN:677588487)*

File: sql/pgwire_internal_test.go [hunk 1]
```diff
@@ -73,14 +73,20 @@ func TestPGWireConnectionCloseReleasesLeases(t *testing.T) {
...
+	leases := ts.active.data
+	ts.mu.Unlock()
+	if len(leases) != 1 {
+		t.Fatalf("expected one lease, found: %d", len(leases))
 	}
 	// Wait for the lease to be released.
 	util.SucceedsSoon(t, func() error {
-		if ts.active.data[0].refcount != 0 {
+		ts.mu.Lock()
+		lease := *(ts.active.data[0])
```
Comment (tamird) [inline sql/pgwire_internal_test.go]: heh - `refcount := ts.active.data[0].refcount`?
anyway, still lgtm

File: sql/pgwire_internal_test.go [hunk 2]
```diff
@@ -73,14 +73,20 @@ func TestPGWireConnectionCloseReleasesLeases(t *testing.T) {
 	if ts == nil {
 		t.Fatal("table state not found")
 	}
-	if len(ts.active.data) != 1 {
-		t.Fatalf("expected one lease, found: %d", len(ts.active.data))
+	ts.mu.Lock()
+	leases := ts.active.data
```
Comment (tamird) [inline sql/pgwire_internal_test.go]: ditto - numLeases := len(...)

**Comment by andreimatei** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[sql/pgwire_internal_test.go, line 85 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7754#-KMQt4q5Eb9mgxdgef8Y-r2-85:-KMRJctP07W1N3PWV6Mv:-192890189) ([raw file](https://github.com/cockroachdb/cockroach/blob/042dddceb119f836462c9a6fab9242aafbec6ab2/sql/pgwire_internal_test.go#L85)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> heh - `refcount := ts.active.data[0].refcount`?
> 
> anyway, still lgtm
</details>
I prefer it light this. Maybe we'll need to read more fields later.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7754)*

**Pull request merged**